File: app/src/main/kotlin/com/openclaw/ghostcrab/domain/repository/GatewayConnectionManager.kt
package com.openclaw.ghostcrab.domain.repository

import com.openclaw.ghostcrab.domain.exception.GatewayAuthException
import com.openclaw.ghostcrab.domain.exception.GatewayTimeoutException
import com.openclaw.ghostcrab.domain.exception.GatewayUnreachableException
import com.openclaw.ghostcrab.domain.model.AuthRequirement
import com.openclaw.ghostcrab.domain.model.GatewayConnection
import kotlinx.coroutines.flow.StateFlow

/**
 * Manages the lifecycle of a single active gateway connection.
 *
 * All state is exposed as a hot [StateFlow]. Only one connection is active at a time.
 *
 * **Contract frozen at v1.0.** Do not change signatures without an ADR.
 */
interface GatewayConnectionManager {

    /**
     * Current connection state. Starts at [GatewayConnection.Disconnected].
     * Updates are delivered on the main thread.
     */
    val connectionState: StateFlow<GatewayConnection>

    /**
     * Determines the authentication requirement for a gateway without storing any state.
     *
     * Probes `GET /health` (unauthenticated) then `GET /status` (unauthenticated).
     * - `/health` fails → [GatewayUnreachableException]
     * - `/status` returns 401 or 403 → [AuthRequirement.Token]
     * - `/status` returns 2xx → [AuthRequirement.None]
     *
     * @param url Base URL of the gateway (scheme + host + port, no trailing slash).
     * @return The [AuthRequirement] inferred from the probe.
     * @throws GatewayUnreachableException if the host is not reachable.
     * @throws GatewayTimeoutException if either probe times out.
     */
    suspend fun probeAuth(url: String): AuthRequirement

    /**
     * Connects to a gateway.
     *
     * 1. Emits [GatewayConnection.Connecting].
     * 2. Calls [probeAuth] to determine auth mode.
     * 3. Instantiates an authenticated Ktor client.
     * 4. Fetches `/status` to read version and capabilities.
     * 5. Emits [GatewayConnection.Connected] on success.
     * 6. On any failure: emits [GatewayConnection.Error] **and** rethrows the exception.
     *
     * Calling [connect] while already [GatewayConnection.Connected] disconnects first.
     *
     * @param url Base URL of the gateway.
     * @param token Bearer token for authentication. Pass `null` for unauthenticated gateways.
     * @throws GatewayUnreachableException if the host is not reachable.
     * @throws GatewayAuthException if authentication fails.
     * @throws GatewayTimeoutException if the connection times out.
     */
    suspend fun connect(url: String, token: String? = null)

    /**
     * Disconnects from the current gateway.
     *
     * Cancels the held Ktor client, emits [GatewayConnection.Disconnected].
     * No-op if already disconnected.
     */
    suspend fun disconnect()
}

File: ios/GhostCrab/Domain/Repository/GatewayConnectionManager.swift
import Foundation

/// Manages the lifecycle of a single active gateway connection.
///
/// All state is exposed as an `AsyncStream<GatewayConnection>`. Only one connection
/// is active at a time.
///
/// **Contract frozen at v1.0.** Do not change signatures without an ADR.
public protocol GatewayConnectionManager: Sendable {

    /// Current connection state. Replays the latest value on subscribe and emits every
    /// subsequent transition. Starts at `GatewayConnection.disconnected`.
    ///
    /// Each call returns an independent stream backed by the same underlying state.
    func connectionStates() -> AsyncStream<GatewayConnection>

    /// Synchronously snapshots the current connection state without subscribing.
    var currentState: GatewayConnection { get }

    /// Determines the authentication requirement for a gateway without storing any state.
    ///
    /// Probes `GET /health` (unauthenticated) then `GET /status` (unauthenticated).
    /// - `/health` fails → throws `GatewayError.unreachable`
    /// - `/status` returns 401 or 403 → `.token`
    /// - `/status` returns 2xx → `.none`
    ///
    /// - Parameter url: Base URL of the gateway (scheme + host + port, no trailing slash).
    /// - Returns: The `AuthRequirement` inferred from the probe.
    /// - Throws: `GatewayError.unreachable` if the host is not reachable.
    /// - Throws: `GatewayError.timeout` if either probe times out.
    func probeAuth(url: String) async throws -> AuthRequirement

    /// Connects to a gateway.
    ///
    /// 1. Emits `GatewayConnection.connecting`.
    /// 2. Calls `probeAuth(url:)` to determine auth mode.
    /// 3. Instantiates an authenticated API client.
    /// 4. Fetches `/status` to read version and capabilities.
    /// 5. Emits `GatewayConnection.connected` on success.
    /// 6. On any failure: emits `GatewayConnection.error` **and** rethrows the error.
    ///
    /// Calling `connect` while already connected disconnects first.
    ///
    /// - Parameters:
    ///   - url: Base URL of the gateway.
    ///   - token: Bearer token for authentication. Pass `nil` for unauthenticated gateways.
    /// - Throws: `GatewayError.unreachable` if the host is not reachable.
    /// - Throws: `GatewayError.auth` if authentication fails.
    /// - Throws: `GatewayError.timeout` if the connection times out.
    func connect(url: String, token: String?) async throws

    /// Disconnects from the current gateway.
    ///
    /// Cancels the held API client, emits `GatewayConnection.disconnected`.
    /// No-op if already disconnected.
    func disconnect() async
}
